src/components/notes/saving-indicator.jsx

```diff
@@ -0,0 +1,83 @@
+import React, { useRef, useState } from 'react'
+import { translate } from 'cozy-ui/react/I18n'
+import { Text } from 'cozy-ui/react/Text'
+
+import styles from 'components/notes/saving-indicator.styl'
+
+// constants
+const sec = 1000
+const min = 60 * sec
+const hour = 60 * min
+const day = 24 * hour
+
+// differents steps for message formating
+// [age, key, unit, coef] :
+// - `age`: the first list where the age of last save
+// if younger than this `age` will be used
+// - `key` is the traduction key for the message
+// - `unit` is the unit in which we will send the time
+// to the translated message
+// - `coef` is  the granularity in which we will send
+// the time (`min, 5` is "by 5 minutes slices")
+const steps = [
+  [10 * sec, 'saved_just_now', sec, 1],
+  [min, 'saved_secs_ago', sec, 10],
+  [2 * min, 'saved_min_ago', min, 1],
+  [5 * min, 'saved_mins_ago', min, 1],
+  [20 * min, 'saved_mins_ago', min, 5],
+  [hour, 'saved_mins_ago', min, 10],
+  [2 * hour, 'saved_hour_ago', hour, 1],
+  [day, 'saved_hours_ago', hour, 1],
+  [2 * day, 'saved_day_ago', day, 1],
+  [9999 * day, 'saved_day_ago', day, 1]
+]
+
+/**
+ * @typedef {object} SavingIndicatorProps
+ * @property {Date} dirtySince - our copy was not in sync since this date
+ * @property {Date} lastSave - last time we managed to save something to the server
+ */
+
+/**
+ * Display  a saving indicator in the bottom right
+ *
+ * @param {SavingIndicatorProps} props
+ */
+const SavingIndicator = translate()(props => {
+  const { dirtySince, lastSave, t } = props
+  const now = new Date()
+
+  const [, setChange] = useState(0)
+  const rerender = () => {
+    setChange(val => val + 1)
+  }
+  const timeout = useRef(undefined)
+
+  if (timeout.current) window.clearTimeout(timeout.current)
```

Comment:
Isn't it some kind of effect? It looks like it should be in a `useEffect`, no?

Reply:
These won't add any side effect in DOM, nor do need to wait until the component is rendered or attached in the DOM. AFAIK `useEffet` is useless here.

Reply:
There is a side effect on window. AMHA the fact that setTimeout is not pure is sufficient to have this in useEffect.

AMHA we should extract all unpure logic outside the component and think of a way to make the component render pure.

Here, I think the intention of setTimeout is not very clear:

Could we have something like 

```
usePeriodicRender(shouldRerender, checkPeriod)
```

encapsulating the window.setTimeout/clearTimeout ?

---

Also I think it would be better to clear the timeout on unmount. I do not think it is done here.

Reply:
Hum… I like that idea :+1: